### PR TITLE
[FLINK-25075] Refactor PlannerExpressionParser, removing the reflection to instantiate it

### PR DESCRIPTION
**Pull request by slinkydeveloper**
## What is the purpose of the change

This PR refactors `PlannerExpressionParser` to be loaded through the `Factory` SPI infrastructure. It also renames `PlannerExpressionParser` to `ExpressionParser`, as the original name is misleading (is it a parser or a planner?!) and it's implicit the fact that the implementation is planner provided, given it lives in the `org.apache.flink.table.delegation` package.

## Brief change log

* Renamed `PlannerExpressionParser` to `ExpressionParser`
* Added a new `ExpressionParserFactory` to load the `ExpressionParser` and added the `DefaultExpressionParserFactory`
* Removed `org.apache.flink.table.expressions.ExpressionParser` as it doesn't provide any particular value 
* Removed `WindowCreationValidationTest` which haven't much sense now. Previously, they were testing the reflection failure with the nice exception, but now we use the `FactoryUtil` methods which are already well tested

Note that the git changelog confuses the `PlannerExpressionParser` renaming with the removal of `org.apache.flink.table.expressions.ExpressionParser`.

## Verifying this change

This change is a trivial rework / code cleanup already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/delegation/ExpressionParser.java [hunk 1]
```diff
@@ -16,27 +16,24 @@
...
- * Parser for expressions inside a String. This parses exactly the same expressions that would be
- * accepted by the Scala Expression DSL.
- *
- * <p>{@link ExpressionParser} use {@link PlannerExpressionParser} to parse expressions.
+ * {@link Expression} parser used by Table API to parse strings in AST expression. This parses
+ * exactly the same expressions that would be accepted by the Scala Expression DSL.
  */
 @Internal
-public final class ExpressionParser {
+public interface ExpressionParser {
```
Comment (twalthr) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/delegation/ExpressionParser.java:31]: deprecate the class

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/delegation/ExpressionParserFactory.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+package org.apache.flink.table.delegation;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.factories.Factory;
+import org.apache.flink.table.factories.FactoryUtil;
+
+/** Factory for {@link ExpressionParser}. */
+@Internal
+public interface ExpressionParserFactory extends Factory {
```
Comment (twalthr) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/delegation/ExpressionParserFactory.java:27]: deprecate the class

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/delegation/ParserImpl.java [hunk 3]
```diff
@@ -64,6 +65,7 @@
     private final Supplier<CalciteParser> calciteParserSupplier;
     private final SqlExprToRexConverterFactory sqlExprToRexConverterFactory;
     private static final ExtendedParser EXTENDED_PARSER = ExtendedParser.INSTANCE;
+    private static final ExpressionParserImpl expressionParser = new ExpressionParserImpl();
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/delegation/ParserImpl.java:68]: why this change? this deprecated parser has nothing to do with the SQL parser.

**Review by twalthr — CHANGES_REQUESTED** [review]
Thanks @slinkydeveloper. I added some comments.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/delegation/ParserImpl.java (hunk 3 above)
Comment (slinkydeveloper) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/delegation/ParserImpl.java:68]: Sorry, that's a leftover of when i tried to move the funcionality inside the `Parser` iface

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/delegation/ExpressionParserFactory.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+package org.apache.flink.table.delegation;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.factories.Factory;
+import org.apache.flink.table.factories.FactoryUtil;
+
+/**
+ * Factory for {@link ExpressionParser}.
+ *
+ * @deprecated The Scala Expression DSL is deprecated
```
Comment (twalthr) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/delegation/ExpressionParserFactory.java:28]: It's actually the `Java String Expression DSL` but I will fix this while merging.

**Review by twalthr — APPROVED** [review]
LGTM
